service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ActionCMOController.java

```diff
@@ -0,0 +1,146 @@
+
+        CaseManagementOrder caseManagementOrderToBeActioned =
+            caseManageOrderActionService.getCaseManagementOrderForAction(caseDataMap);
+        caseDetails.getData().put(CASE_MANAGEMENT_ORDER_KEY, caseManagementOrderToBeActioned);
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(
+        @RequestHeader(value = "authorization") String authorization,
+        @RequestHeader(value = "user-id") String userId,
+        @RequestBody CallbackRequest callbackRequest) throws IOException {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        Map<String, Object> caseDataMap = caseDetails.getData();
+
+        CaseManagementOrder caseManagementOrderToBeActioned =
+            caseManageOrderActionService.getCaseManagementOrderForAction(caseDataMap);
+
+        Document updatedDraftCMODocument = getDocument(authorization, userId, caseDataMap, false);
+        CaseManagementOrder updatedDraftCaseManagementOrderWithDocument =
+            caseManageOrderActionService.addDocumentToCaseManagementOrder(caseManagementOrderToBeActioned,
+                updatedDraftCMODocument);
+
+        caseDetails.getData().put(CASE_MANAGEMENT_ORDER_ACTION_KEY, ImmutableMap.of("orderDoc",
+            updatedDraftCaseManagementOrderWithDocument.getOrderDoc()));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(
+        @RequestHeader(value = "authorization") String authorization,
+        @RequestHeader(value = "user-id") String userId,
+        @RequestBody CallbackRequest callbackRequest) throws IOException {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+
+        CaseManagementOrder updatedDraftCaseManagementOrder =
+            caseManageOrderActionService.getCaseManagementOrderForAction(caseDetails.getData());
+
+        boolean judgeApprovedDraftCMO = hasJudgeApprovedDraftCMO(updatedDraftCaseManagementOrder);
+        Document actionedCaseManageOrderDocument = getDocument(authorization, userId, caseDetails.getData(),
+            judgeApprovedDraftCMO);
+
+        CaseManagementOrder updatedCaseManagementOrderWithDocument =
+            caseManageOrderActionService.addDocumentToCaseManagementOrder(updatedDraftCaseManagementOrder,
+                actionedCaseManageOrderDocument);
+
+        CaseManagementOrderAction caseManagementOrderAction =
+            updatedCaseManagementOrderWithDocument.getCaseManagementOrderAction();
+
+        prepareCaseDetailsForSubmission(caseDetails, caseManagementOrderAction, updatedCaseManagementOrderWithDocument,
+            judgeApprovedDraftCMO);
```

Comment:
I find this pretty confusing. Could we not have a void method that adds the document to your original cmo object?